plugins/woocommerce/includes/class-wc-cart-session.php

```diff
@@ -207,7 +207,20 @@ public function get_cart_from_session() {
 					)
 				);
 
+				/**
+				 * Filter to modify or add session data to the cart contents.
+				 *
+				 * @since 3.2.0
+				 *
+				 * @param array  $session_data Data for an item in the cart.
+				 * @param array  $values       Data for an item in the cart, without the product object.
+				 * @param string $key          The cart item hash.
+				 */
 				$cart_contents[ $key ] = apply_filters( 'woocommerce_get_cart_item_from_session', $session_data, $values, $key );
+				if ( ! isset( $cart_contents[ $key ]['data'] ) || ! $cart_contents[ $key ]['data'] instanceof WC_Product ) {
+					// If the cart contents is missing the product object after filtering, add it back in.
+					$cart_contents[ $key ]['data'] = $product;
+				}
```

Comment:
>I'm not sure what other logging would be useful here, but I'm open to ideas.

I mean, if we can add log a warning something like,

`Seems like a hook on "woocommerce_get_cart_item_from_session" is changing the type of data to something other than `WC_Product`, we have added the product back to the cart content....`

Something on these lines, to make folks aware.

This is something from top of my head, maybe I am being extra cautious.

Reply:
Fair enough, added a `doing_it_wrong` in 761f4f53a0. That way it will go to the error log, but only if they have WP_DEBUG defined, so it shouldn't get noisy in production environments. Wdyt?

Reply:
Thanks for adding it, seems like an apt middle ground. If we get complaints, we can iterate on this, but what its worth, it is much better than a fatal. Thanks for your work on this.